src/main/java/dev/latvian/mods/kubejs/core/mixin/RecipeManagerMixin.java

```diff
@@ -31,11 +43,72 @@ public abstract class RecipeManagerMixin implements RecipeManagerKJS {
 	@Unique
 	private ReloadableServerResourcesKJS kjs$resources;
 
-	@Inject(method = "apply(Ljava/util/Map;Lnet/minecraft/server/packs/resources/ResourceManager;Lnet/minecraft/util/profiling/ProfilerFiller;)V", at = @At("HEAD"), cancellable = true)
+	@Unique
+	private RecipesKubeEvent kjs$event;
+
+	@Inject(
+		method = "apply(Ljava/util/Map;Lnet/minecraft/server/packs/resources/ResourceManager;Lnet/minecraft/util/profiling/ProfilerFiller;)V",
+		at = @At("HEAD")
+	)
 	private void customRecipesHead(Map<ResourceLocation, JsonElement> map, ResourceManager resourceManager, ProfilerFiller profiler, CallbackInfo ci) {
-		if (kjs$resources.kjs$getServerScriptManager().recipes(this, resourceManager, map)) {
-			ci.cancel();
+		var manager = kjs$resources.kjs$getServerScriptManager();
+
+		// FIXME: data maps!
+		if (ServerEvents.COMPOSTABLE_RECIPES.hasListeners()) {
+			ServerEvents.COMPOSTABLE_RECIPES.post(ScriptType.SERVER, new CompostableRecipesKubeEvent());
+		}
+
+		for (var entry : manager.getRegistries().cachedRegistryTags.values()) {
+			if (entry.registry() == null || entry.lookup() == null) {
+				continue;
+			}
+
+			entry.registry().bindTags(Cast.to(entry.lookup().bindingMap()));
 		}
+
+		manager.recipeSchemaStorage.fireEvents(manager.getRegistries(), resourceManager);
+
+		SpecialRecipeSerializerManager.INSTANCE.reset();
+		ServerEvents.SPECIAL_RECIPES.post(ScriptType.SERVER, SpecialRecipeSerializerManager.INSTANCE);
+
+		if (ServerEvents.RECIPES.hasListeners()) {
+			ConsoleJS.SERVER.info("Processing recipes...");
+			kjs$event = new RecipesKubeEvent(manager, resourceManager);
+			kjs$event.post(this, map);
+		}
+	}
+
+	@Inject(
+		method = "apply(Ljava/util/Map;Lnet/minecraft/server/packs/resources/ResourceManager;Lnet/minecraft/util/profiling/ProfilerFiller;)V",
+		at = @At(value = "INVOKE", target = "Lorg/slf4j/Logger;error(Ljava/lang/String;Ljava/lang/Object;Ljava/lang/Object;)V")
+	)
+	private void catchFailingRecipes(CallbackInfo ci, @Local Map.Entry<ResourceLocation, JsonElement> entry, @Local RuntimeException ex) {
+		if (kjs$event != null) {
+			kjs$event.handleFailedRecipe(entry.getKey(), entry.getValue(), ex);
+		}
+	}
+
+	@Inject(
+		method = "apply(Ljava/util/Map;Lnet/minecraft/server/packs/resources/ResourceManager;Lnet/minecraft/util/profiling/ProfilerFiller;)V",
+		at = @At(value = "TAIL")
+	)
+	@SuppressWarnings("removal")
+	private void addServerData(CallbackInfo ci) {
+		if (kjs$event != null) {
+			// FIXME: please remove this soon! massive performance implications!
+			var recipesByName = new HashMap<>(byName);
+
+			KubeJSPlugins.forEachPlugin(p -> p.injectRuntimeRecipes(kjs$event, this, recipesByName));
+
+			kjs$event.finishEvent();
+
+			// make sure byType is also set correctly
+			kjs$replaceRecipes(recipesByName);
+
+			kjs$getResources().kjs$getServerScriptManager().serverData = new SyncServerDataPayload(KubeServerData.collect());
```

Comment:
Now that this heavily depends on being listened to Recipe Events, this now breaks because it is null when we don't listen to Recipe Events:
https://github.com/KubeJS-Mods/KubeJS/blob/c0f67bf14c8f722709d714af5dd29608329bb839/src/main/java/dev/latvian/mods/kubejs/player/KubeJSPlayerEventHandler.java#L28

Also, is it intended that modify tooltip event should only work if listening to Recipe Events?

This is only called if listening to recipe events as:
1. `kjs$event` needs to be non-null
2. `KubeServerData.collect()` is called
3. `ItemEvents.MODIFY_TOOLTIPS#post` is finally called (this is the only reference to this post)
https://github.com/KubeJS-Mods/KubeJS/blob/c0f67bf14c8f722709d714af5dd29608329bb839/src/main/java/dev/latvian/mods/kubejs/net/KubeServerData.java#L26-L28